### ISSUE TITLE
Model Optimizations

### ISSUE THREAD

**Issue by AmitMY** [sign-language-processing/sign-vq#2]
- [x] Ideally, the `step` operation should be compiled with `torch.compile`
https://github.com/sign-language-processing/sign-vq/blob/main/sign_vq/model.py#L204-L219

But probably, it would only be possible to compile separate functions, such as the loss (https://github.com/sign-language-processing/sign-vq/blob/main/sign_vq/model.py#L130-L149) and the forward (https://github.com/sign-language-processing/sign-vq/blob/main/sign_vq/model.py#L164-L165)

----

- [ ] Step should also be wrapped with autocast to `bfloat16` probably
https://github.com/Lightning-AI/pytorch-lightning/issues/19980

----

- [ ] As far as ugly numbers go - we use `pose_dims: tuple = (178, 3),` which is 534, a number that doesn't even divide by 4.
Since the `channels` here are irrelevant for powers of 2 (3 is a prime number), we might want to pad our inputs to be divisible by 16 (in this case, 192). We might see performance improvements from that

----

- [x] Gradient clipping (grad norm never above 1, for example), might be useful for the training.

----

- [ ] We should take another look at the gradient warmup
https://github.com/sign-language-processing/sign-vq/blob/main/sign_vq/model.py#L167-L193

Specifically, we set `warmup_steps` to 10,000 which is only 400, and we would like better control over it.

----

- [x] AdamW should run with `fused=true`, since it is faster.
Only when `'fused' in inspect.signature(torch.optim.AdamW).parameters and 'cuda' in device`

----

- [ ] Use a larger batch size. Understand how much we currently use per GPU, and ideally, running on multiple GPUs `DistributedDataParallel`

```
    # TODO: https://lightning.ai/docs/pytorch/stable/advanced/training_tricks.html requiring LightningDataModule
    # tuner = Tuner(trainer)
    # # Auto-scale batch size with binary search
    # tuner.scale_batch_size(model, mode="binsearch")
```

---- 

- [x] In every batch, we take a number of poses, with [`max_length=512`](https://github.com/sign-language-processing/sign-vq/blob/main/sign_vq/dataset.py#L49)
However, most poses in our dataset are shorter than 512. 

Packing is required.

Therefore, perhaps more ideally, all poses should be concatenated to a single numpy array (or shards of number of frames), with perhaps some special frames in between for padding (for example, 10 frames of zeros, that do not take part in the loss calculation) OR better a "id" tensor that has a unique number for every video `[1,1,1,2,2,2,2,3,3,3,3]`, then we can use it to mask the self-attention in the encoder and decoder.
Then, the data loader fully loads a shard, and iterates it sequentially, 512 frames per item in a batch.

This is not perfect. Ideally, we want to "shuffle" the data every epoch. To do that, we can load all the data from the zip file, and every epoch randomly concatenate it differently (possibly in an online fashion). The larger the data becomes, the less this is a problem.

Bonus: this will make sure all positional embeddings learn equally~
Right now we use sinusodial embeddings (https://github.com/sign-language-processing/sign-vq/blob/main/sign_vq/model.py#L173) because of fear that they won't learn equally. 

----

- [ ] Assuming data like above is now stringed together, it makes sense to support larger context sizes.
We should aim for at least 30 seconds at 30fps, or 900 frames. For simplicity, 1024 seems like a good number.

For extra usefulness for this model as a generic pose encoder, a [special] token can be added in the beginning, to learn a generic representation for the pose (does not take part in loss calculation)

----

- [ ] Currently, the model needs to assign information for both the appearance and the movement of the signer. 
We could use something like https://github.com/sign-language-processing/pose-anonymization to remove the appearance, then learn only on the anonymized poses (that also makes all vectors closer to zero), and only relative movement is captured around zero
